### PR TITLE
Added docs on branching and source control

### PR DESCRIPTION
**Pull request by trmartin4**
## Objective

I moved the Branching Strategy docs from [Confluence](https://bitwarden.atlassian.net/wiki/spaces/EN/pages/231211021/Branching+Strategy) and combined them with the existing Long-Lived Feature Branch page.

Moved branching and PR-related content from [Git Best Practices](https://bitwarden.atlassian.net/wiki/spaces/EN/pages/223543674/Git+Best+Practices) with the goal of deleting that page from Confluence as well.

### CODE REVIEW

**Review by differsthecat — COMMENTED** [review]
Just thinking through this, is this something we want to have here in the contributing docs? It seems to be very specific to internal processes and I'm not sure what the line is between having stuff here in the docs vs. in Confluence.

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [inline docs/contributing/pull-requests/index.md:109]: Do we want to look into adding a group for the design people in each pod? Or keep it as direct references by the pod?

File: docs/contributing/pull-requests/index.md
Comment (Hinton) [inline docs/contributing/pull-requests/index.md:69]: We could move this up into the previous section. Do you mind updating line 62 to follow the convention?

`git checkout -b <pod>/<jira-tag> -t origin/master`

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:36]: Should we just remove `feature` from this document? I think _Long-Lived Branch_ works just as well as _Long-Lived Feature Branch_

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:73]: This section seems a bit repetitive of the _Which Branching Model to Choose?_ section.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:128]: We can remove the bitwarden tag since this whole document is now internal.

```suggestion
We typically denote one person as being responsible for keeping the Long-Lived Feature Branch up to
date. In most cases this will be the TPC of the pod but can be anyone. This
person is responsible for keeping the feature branch reasonably up to date with master, preparing it
for merge and more.
```

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:184]: Should we just remove the `optional-description` mentioned in the branching strategy since we seem to discourage it.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:193]: and design if it requires approval from design?

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:196]: We seem to follow this for long lived branches as well. Remove?

**Comment by trmartin4** [comment]
> @differsthecat I think it's beneficial to keep documentation in a single place to avoid jumping around to much. Although the same or similar could be achieved by having a link to confluence.

@Hinton do you have a preference of which to do?  My thinking was that since the "Long-Lived Feature Branch" documentation was here, we needed to pull in the rest of it (the "Short-Lived" version).  I am fine moving back out to Confluence, but I think we should move out all of the branching docs as a whole and reference them in Confluence.

On the one hand, this branching bleeds into our SDLC quite a bit, and that _is_ in Confluence.  On the other hand, I think it's nice to have all of the docs to do development in one place.

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:196]: Do we?  On the SG pod, when an individual developer finishes their task that is part of a feature (I'll call this the "issue branch"), they complete the PR to merge that change into the feature branch, and QA tests on the feature branch.  This was our operating model as it allows QA to keep their environments pointed to one place instead of switching around.

I do see your point that if we were to test **before** merging it would avoid the mental math that you have to do every time you have a PR approved.  (i.e. "Is this merging into a feature branch, so I complete it?  Or is it merging into `master` so I leave it open for testing?")

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:36]: I've been in several discussions that indicate we might be moving toward adopting a more "by the books" gitflow branching strategy. That's why I kept the "feature" nomenclature.  Also, one of the things I've heard from new engineers coming in is that our slightly different branching strategy takes some time to understand, so I was trying to build in the comfort of knowing that these can be thought of as "feature branches" in the terminology of gitflow.

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:184]: Done!

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:69]: By "section", I think you meant set of <bitwarden> tags, correct?  If so, I've moved it.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:109]: I've reached out to Danielle to get her opinion on this.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [reply docs/contributing/pull-requests/branching.md:196]: Ah now I get it. I read it as the long lived feature branch could be merged to master. Can we rephrase it to make it more clear that we mean merging into a feature branch, not merging the feature branch?

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [reply docs/contributing/pull-requests/branching.md:36]: That's fair. Feature branches is a well known term.

**Review by Hinton — APPROVED** [review]
I think this looks good. And I'm happy keeping it in the contributing docs.

File: docs/contributing/pull-requests/index.md
Comment (trmartin4) [reply docs/contributing/pull-requests/index.md:105]: They already have the `dept-design` tag, so I added that to the docs.

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:73]: I clarified this, as it was very repetitive.

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:196]: I clarified this.  It was confusing and could be read in different ways.

**Comment by trmartin4** [comment]
> I think this looks good. And I'm happy keeping it in the contributing docs.

I made some more changes to address the comments you made, as I didn't want them to go un-addressed into `master`.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:68]: I think the term `repository` is slightly odd here and could be a bit confusing in git terminology.

```suggestion
A Long-Lived Feature Branch describes a branch that serves as a repository for the changes required
```

**Comment by Hinton** [comment]
With the move to a single jira project should we update the branch naming page?

File: docs/contributing/pull-requests/branching.md
Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md]: That's a great point.  Repo is a loaded term here.  I've restructured that sentence.

**Comment by trmartin4** [comment]
> With the move to a single jira project should we update the branch naming page?

I've updated the Naming section as well as all branch references on the page.

File: docs/contributing/pull-requests/branching.md
Comment (Hinton) [inline docs/contributing/pull-requests/branching.md:18]: This is a personal preference, but I actually find the team prefix to provide more value now that (almost) everyone will use the same jira tag. It's going to be quite difficult tie branches to teams otherwise.

Comment (trmartin4) [reply docs/contributing/pull-requests/branching.md:18]: I had the same thought, but without Jira's built-in abbreviations, we would need to agree upon them as a team and stay fairly consistent.  I can propose a prefix for each team on the docs if you'd like.  Would you prefer an abbreviation (Auth) or an initialism like Jira uses (AT or SM)?  Or is a combination of both OK?

**Review by Hinton — APPROVED** [review]

**Pull request merged**
